lib/fog/openstack.rb

```diff
@@ -331,17 +330,96 @@ def self.authenticate_v3(options, connection_options = {})
-          :current_user_id          => body['token']['user']['id'],
-          :unscoped_token           => options[:unscoped_token]
+        :user => body['token']['user']['name'],
+        :tenant => tenant,
+        :identity_public_endpoint => identity_url,
+        :server_management_url => management_url,
+        :token => token,
+        :expires => body['token']['expires_at'],
+        :current_user_id => body['token']['user']['id'],
+        :unscoped_token => options[:unscoped_token]
       }
     end
 
+    #
+    # Try to validate a token before re-authenticate to avoid token sprawl inside keystone
+    # If validation fails, use old re-authentication logic to be compliant
+    #
+    def self.validate_or_authenticate_v3(options, connection_options)
+      if options[:openstack_auth_token].nil? || options[:openstack_auth_token].empty?
+        return authenticate_v3(options, connection_options)
+      else
+        ret = nil
+        begin
+          ret = validate_v3(options, connection_options)
+          unless ret
+            ret = authenticate_v3(options, connection_options)
+          end
+        rescue
+          ret = authenticate_v3(options, connection_options)
+        end
+        return ret
+      end
+    end
+
+    #
+    # Validate a token
+    #
+    def self.validate_v3(options, connection_options)
+      token = options[:openstack_auth_token]
+      uri = options[:openstack_auth_uri]
+      omit_default_port = options[:openstack_auth_omit_default_port]
+
+      path = (uri.path && !uri.path.empty?) ? uri.path : 'v3'
+
+      connection = Fog::Core::Connection.new(uri.to_s, false, connection_options)
+      request = {
+        :expects => [200],
+        :headers => {'Content-Type' => 'application/json', 'X-Auth-Token' => token, 'X-Subject-Token' => token},
+        :method => 'GET',
+        :path => path
+      }
+      request[:omit_default_port] = omit_default_port unless omit_default_port.nil?
+      response = connection.request(request)
+      body = Fog::JSON.decode(response.body)
+
+      if body['token']['project']
+        tenant = body['token']['project']
+      elsif body['token']['user']['project']
+        tenant = body['token']['user']['project']
+      end
+
+      service_type = options[:openstack_service_type]
+      service_name = options[:openstack_service_name]
+      endpoint_type = map_endpoint_type(options[:openstack_endpoint_type] || 'publicURL')
+
+      identity_service_type = options[:openstack_identity_service_type]
+      openstack_region = options[:openstack_region]
+
+      service = get_service_v3(body, service_type, service_name, openstack_region, options)
+      identity_service = get_service_v3(body, identity_service_type, nil, nil, :openstack_endpoint_path_matches => %r{/v3}) if identity_service_type
+
+      # Services not found  -> unscoped token
+      # Fallback to standard authenticate to be compatible
+      if service
```

Comment:
Could the service validation be in its own method to separate it from the token validation itself?
Although, when OpenStack returns a token there are some services attach to it through the catalog. Normally that's what should be used for service validation, now if that doesn't fit fog-openstack it's a different story.